crates/webidl/src/first_pass.rs

```diff
@@ -103,43 +114,177 @@ impl FirstPass for webidl::ast::Enum {
     }
 }
 
-impl FirstPass for webidl::ast::Interface {
-    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>) -> Result<()> {
+impl FirstPass<()> for webidl::ast::Interface {
+    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>, (): ()) -> Result<()> {
         use webidl::ast::Interface::*;
 
         match self {
-            NonPartial(interface) => interface.first_pass(record),
-            _ => {
-                // Other interfaces aren't currently used in the first pass
+            Partial(interface) => interface.first_pass(record, ()),
+            NonPartial(interface) => interface.first_pass(record, ()),
+            // TODO
+            Callback(..) => {
+                warn!("Unsupported WebIDL interface: {:?}", self);
                 Ok(())
             }
         }
     }
 }
 
-impl FirstPass for webidl::ast::NonPartialInterface {
-    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>) -> Result<()> {
-        if record.interfaces.insert(self.name.clone()) {
-            warn!("Encountered multiple declarations of {}", self.name);
+impl FirstPass<()> for webidl::ast::NonPartialInterface {
+    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>, (): ()) -> Result<()> {
+        if record.interfaces.contains_key(&self.name) {
+            if let Some(interface_data) = record.interfaces.get_mut(&self.name) {
+                if interface_data.partial {
+                    interface_data.partial = false;
+                } else {
+                    warn!("Encountered multiple declarations of {}", self.name);
+                }
+            }
+        } else {
+            record.interfaces.insert(
+                self.name.clone(),
+                InterfaceData {
+                    partial: false,
+                    operations: Default::default(),
+                    overloaded_operations: Default::default(),
+                    has_constructor: false,
+                    has_overloaded_constructors: false,
+                },
+            );
+        }
+
+        if ::util::is_chrome_only(&self.extended_attributes) {
+            return Ok(())
+        }
+
+        for extended_attribute in &self.extended_attributes {
+            extended_attribute.first_pass(record, &self.name)?;
+        }
+
+        for member in &self.members {
+            member.first_pass(record, &self.name)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl FirstPass<()> for webidl::ast::PartialInterface {
+    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>, (): ()) -> Result<()> {
+        if !record.interfaces.contains_key(&self.name) {
```

Comment:
Similar to above this may look a bit cleaner with the `entry` method, but if this shows up in a few places it could also be a helper method perhaps?